File: mobile/ios/Sources/ReviewService.swift
import Foundation

/// Networking for guest reviews against the live Next.js API. Mirrors
/// `BookingService`: pure URLSession + Codable, reading the bearer token from
/// `UserDefaults` under `AuthStore.tokenKey` ("qk_token").
///
///   GET  {base}/api/local/reviews?listing_id=ID   (public)
///         → [{ rating, comment, reviewer_name, created_at }]
///   GET  {base}/api/local/reviews                  (Bearer)
///         → reviewable stays (confirmed, past checkout, not yet reviewed)
///   POST {base}/api/local/reviews { booking_id, rating, comment }  (Bearer)
///         → 201 (success)
struct ReviewService {
    static let shared = ReviewService()

    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        cfg.waitsForConnectivity = true
        return URLSession(configuration: cfg)
    }()

    /// The persisted bearer token, or `nil` when browsing as a guest.
    var token: String? {
        let value = UserDefaults.standard.string(forKey: AuthStore.tokenKey)
        return (value?.isEmpty == false) ? value : nil
    }

    // MARK: - Public reviews for a listing

    /// Reviews for a listing (public — no auth required). Returns an empty list
    /// on a non-2xx so the detail screen simply shows no reviews.
    func fetchReviews(listingID: String) async throws -> [Review] {
        var components = URLComponents(string: "\(Config.apiBaseURL)/api/local/reviews")!
        components.queryItems = [URLQueryItem(name: "listing_id", value: listingID)]
        let url = components.url!

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ReviewError.message("Invalid response from the server.")
        }
        guard (200...299).contains(http.statusCode) else {
            throw ReviewError.message(Self.decodeError(data) ?? "Couldn't load reviews (\(http.statusCode)).")
        }
        return try JSONDecoder().decode([Review].self, from: data)
    }

    // MARK: - Reviewable stays (signed-in)

    /// The signed-in user's stays that are eligible for a review (confirmed,
    /// past checkout, not yet reviewed). Throws `ReviewError.notSignedIn` when
    /// there is no token or the server returns 401.
    func fetchReviewable() async throws -> [ReviewableStay] {
        guard let token else { throw ReviewError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/reviews")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ReviewError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ReviewError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ReviewError.message(Self.decodeError(data) ?? "Couldn't load reviewable stays (\(http.statusCode)).")
        }
        return try JSONDecoder().decode([ReviewableStay].self, from: data)
    }

    /// Whether the given booking is still awaiting a review (best-effort; returns
    /// `false` on any error). Used to decide whether to surface the "Leave a
    /// review" entry on a reservation.
    func isReviewable(bookingID: String) async -> Bool {
        guard let stays = try? await fetchReviewable() else { return false }
        return stays.contains { $0.bookingId == bookingID }
    }

    // MARK: - Submit a review

    /// Submit a 1–5 star review (+ optional comment + up to 6 photos) for a
    /// completed booking. `photos` is an array of `data:image/*` or `http(s)`
    /// image URL strings. Throws `ReviewError.notSignedIn` with no token, or
    /// `ReviewError.message` carrying the server's `{ error }` for 4xx responses.
    func submit(bookingID: String, rating: Int, comment: String, photos: [String] = []) async throws {
        guard let token else { throw ReviewError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/reviews")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "booking_id": bookingID,
            "rating": rating,
            "comment": comment,
            "photos": Array(photos.prefix(6)),
        ])

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ReviewError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ReviewError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ReviewError.message(Self.decodeError(data) ?? "Couldn't submit your review (\(http.statusCode)).")
        }
    }

    // MARK: - Guest reviews (host → guest)

    /// Reviews left about a guest (public — no auth required). Returns an empty
    /// list on a non-2xx so a profile simply shows no guest reviews.
    func fetchGuestReviews(guestID: String) async throws -> [GuestReview] {
        var components = URLComponents(string: "\(Config.apiBaseURL)/api/local/guest-reviews")!
        components.queryItems = [URLQueryItem(name: "guest_id", value: guestID)]
        let url = components.url!

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ReviewError.message("Invalid response from the server.")
        }
        guard (200...299).contains(http.statusCode) else {
            throw ReviewError.message(Self.decodeError(data) ?? "Couldn't load reviews (\(http.statusCode)).")
        }
        return try JSONDecoder().decode([GuestReview].self, from: data)
    }

    /// The signed-in host's past guests that are eligible for a guest review
    /// (confirmed, past checkout, not yet reviewed by the host). Throws
    /// `ReviewError.notSignedIn` when there is no token or the server returns 401.
    func fetchReviewableGuests() async throws -> [ReviewableGuest] {
        guard let token else { throw ReviewError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/guest-reviews")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ReviewError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ReviewError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ReviewError.message(Self.decodeError(data) ?? "Couldn't load reviewable guests (\(http.statusCode)).")
        }
        return try JSONDecoder().decode([ReviewableGuest].self, from: data)
    }

    /// Host leaves (or replaces) a 1–5 star review of a past guest for a booking.
    /// Throws `ReviewError.notSignedIn` with no token, or `ReviewError.message`
    /// carrying the server's `{ error }` for 4xx responses.
    func submitGuestReview(bookingID: String, rating: Int, comment: String) async throws {
        guard let token else { throw ReviewError.notSignedIn }

        let url = URL(string: "\(Config.apiBaseURL)/api/local/guest-reviews")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "booking_id": bookingID,
            "rating": rating,
            "comment": comment,
        ])

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ReviewError.message("Invalid response from the server.")
        }
        if http.statusCode == 401 { throw ReviewError.notSignedIn }
        guard (200...299).contains(http.statusCode) else {
            throw ReviewError.message(Self.decodeError(data) ?? "Couldn't submit your review (\(http.statusCode)).")
        }
    }

    // MARK: - Helpers

    private static func decodeError(_ data: Data) -> String? {
        struct ErrorBody: Decodable { let error: String }
        return (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
    }
}

/// Errors surfaced to the reviews UI.
enum ReviewError: LocalizedError {
    case notSignedIn
    case message(String)

    var errorDescription: String? {
        switch self {
        case .notSignedIn: return "Sign in to leave a review"
        case let .message(text): return text
        }
    }
}

File: mobile/android/app/src/main/java/com/quickin/app/ReviewService.kt
package com.quickin.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * Minimal HTTP client for the local Next.js **reviews** API. Mirrors [BookingService]:
 * no third-party HTTP/JSON libraries (HttpURLConnection + org.json on Dispatchers.IO).
 *
 *   GET  {base}/api/local/reviews?listing_id=ID          (public) -> Review[]
 *   GET  {base}/api/local/reviews                         (auth)   -> ReviewableStay[]
 *   POST {base}/api/local/reviews { booking_id, rating, comment } (auth) -> 201
 */
object ReviewService {

    /** Thrown so callers can distinguish "sign in" (401) from validation (400). */
    class HttpError(val code: Int, message: String) : RuntimeException(message)

    /** Public list of a listing's guest reviews (newest-first). No auth required. */
    suspend fun fetchListingReviews(listingId: String): List<Review> = withContext(Dispatchers.IO) {
        val q = URLEncoder.encode(listingId, "UTF-8")
        val text = getPublic("/api/local/reviews?listing_id=$q")
        parseReviews(text)
    }

    /**
     * Stays the signed-in user can review (confirmed, past checkout, not yet reviewed).
     * Throws [HttpError] (401 when signed out).
     */
    suspend fun fetchReviewableStays(token: String): List<ReviewableStay> =
        withContext(Dispatchers.IO) {
            val text = get(token, "/api/local/reviews")
            parseReviewableStays(text)
        }

    /**
     * Submits a review for a completed [bookingId] (`POST /api/local/reviews`). [rating] is 1–5.
     * [photos] are `data:image/…` or `http(s)` URL strings (≤6 sent). Throws [HttpError]
     * (401 not signed in, 400 already reviewed / not eligible).
     */
    suspend fun submitReview(
        token: String,
        bookingId: String,
        rating: Int,
        comment: String?,
        photos: List<String> = emptyList()
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("booking_id", bookingId)
            put("rating", rating.coerceIn(1, 5))
            if (!comment.isNullOrBlank()) put("comment", comment.trim())
            put("photos", JSONArray(photos.take(6)))
        }
        send("POST", token, "/api/local/reviews", body)
        Unit
    }

    // ---- Guest reviews (host → guest) -----------------------------------------

    /** Public list of reviews left about a guest (newest-first). No auth required. */
    suspend fun fetchGuestReviews(guestId: String): List<GuestReview> = withContext(Dispatchers.IO) {
        val q = URLEncoder.encode(guestId, "UTF-8")
        val text = getPublic("/api/local/guest-reviews?guest_id=$q")
        parseGuestReviews(text)
    }

    /**
     * Past guests the signed-in host can still review (completed stays, not yet reviewed).
     * Throws [HttpError] (401 when signed out / not a host).
     */
    suspend fun fetchReviewableGuests(token: String): List<ReviewableGuest> =
        withContext(Dispatchers.IO) {
            val text = get(token, "/api/local/guest-reviews")
            parseReviewableGuests(text)
        }

    /**
     * Submits a host's review of the guest on a completed [bookingId]
     * (`POST /api/local/guest-reviews`). [rating] is 1–5. Throws [HttpError] (401/400).
     */
    suspend fun submitGuestReview(
        token: String,
        bookingId: String,
        rating: Int,
        comment: String?
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("booking_id", bookingId)
            put("rating", rating.coerceIn(1, 5))
            if (!comment.isNullOrBlank()) put("comment", comment.trim())
        }
        send("POST", token, "/api/local/guest-reviews", body)
        Unit
    }

    // ---- HTTP helpers (mirror BookingService) ---------------------------------

    private fun getPublic(path: String): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
        }
        try {
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun get(token: String, path: String): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun send(method: String, token: String, path: String, body: JSONObject): String {
        val conn = (URL("${Config.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", "Bearer $token")
        }
        try {
            conn.outputStream.use { out -> out.write(body.toString().toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val text = readBody(conn, code)
            if (code !in 200..299) throw HttpError(code, extractError(text, code))
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun readBody(conn: HttpURLConnection, code: Int): String {
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        return stream?.bufferedReader()?.use { it.readText() }.orEmpty()
    }

    private fun extractError(text: String, code: Int): String {
        val parsed = runCatching { JSONObject(text).optString("error") }.getOrNull()
        return if (!parsed.isNullOrBlank()) parsed else "Request failed ($code)"
    }

    // ---- Parsing --------------------------------------------------------------

    private fun parseReviews(json: String): List<Review> {
        val arr = JSONArray(json)
        val out = ArrayList<Review>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            out.add(
                Review(
                    rating = o.optInt("rating", 0),
                    comment = o.optString("comment").ifBlank { null },
                    reviewerName = o.optString("reviewer_name").ifBlank { null },
                    createdAt = o.optString("created_at").ifBlank { null },
                    photos = parseStringArray(o.opt("photos"))
                )
            )
        }
        return out
    }

    /** Coerces a "photos" value (JSON array, or null/absent) into a list of non-blank URL strings. */
    private fun parseStringArray(value: Any?): List<String> {
        val arr = value as? JSONArray ?: return emptyList()
        val out = ArrayList<String>(arr.length())
        for (i in 0 until arr.length()) {
            val s = arr.optString(i).trim()
            if (s.isNotEmpty()) out.add(s)
        }
        return out
    }

    private fun parseGuestReviews(json: String): List<GuestReview> {
        val arr = JSONArray(json)
        val out = ArrayList<GuestReview>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            out.add(
                GuestReview(
                    id = o.optString("id"),
                    bookingId = o.optString("booking_id").ifBlank { null },
                    guestId = o.optString("guest_id").ifBlank { null },
                    hostId = o.optString("host_id").ifBlank { null },
                    rating = o.optInt("rating", 0),
                    comment = o.optString("comment").ifBlank { null },
                    createdAt = o.optString("created_at").ifBlank { null },
                    hostName = o.optString("host_name").ifBlank { null }
                )
            )
        }
        return out
    }

    private fun parseReviewableGuests(json: String): List<ReviewableGuest> {
        val arr = JSONArray(json)
        val out = ArrayList<ReviewableGuest>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            val bookingId = o.optString("booking_id").ifBlank { o.optString("id") }
            if (bookingId.isBlank()) continue
            out.add(
                ReviewableGuest(
                    bookingId = bookingId,
                    listingId = o.optString("listing_id").ifBlank { null },
                    title = o.optString("title").ifBlank { "Your listing" },
                    guestName = o.optString("guest_name").ifBlank { null },
                    checkOut = o.optString("check_out").ifBlank { null }
                )
            )
        }
        return out
    }

    private fun parseReviewableStays(json: String): List<ReviewableStay> {
        val arr = JSONArray(json)
        val out = ArrayList<ReviewableStay>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            // The id used to POST a review is the booking id; tolerate either key.
            val bookingId = o.optString("booking_id").ifBlank { o.optString("id") }
            if (bookingId.isBlank()) continue
            out.add(
                ReviewableStay(
                    bookingId = bookingId,
                    listingId = o.optString("listing_id").ifBlank { null },
                    title = o.optString("title").ifBlank { "Your stay" },
                    location = o.optString("location").ifBlank { null },
                    image = o.optString("image").ifBlank { null },
                    checkIn = o.optString("check_in").ifBlank { null },
                    checkOut = o.optString("check_out").ifBlank { null }
                )
            )
        }
        return out
    }
}
